dataverse/webapi/C#-NETx/BulkOperations/UpsertMultiple/Program.cs

```diff
@@ -0,0 +1,349 @@
+                                    $"{tableLogicalName}s({tableAlternateKey}='sample pre-upsert record key {i + 1:0000000}')"));
+                            }
+                        }
+                    }
+
+                    // Use UpsertMultiple
+                    UpsertMultipleRequest upsertMultipleRequest = new(
+                        entitySetName: tableSetName,
+                        targets: entityList);
+
+                    // Add a tag optional parameter to set a shared variable to be available to a plug-in.
+                    // See https://learn.microsoft.com/power-apps/developer/data-platform/optional-parameters?tabs=webapi#add-a-shared-variable-to-the-plugin-execution-context
+
+                    // TODO?: We could add a string 'tag' parameter to the UpsertMultipleRequest constructor and append it to the URL when it is provided.
+                    // CreateMultiple and UpdateMultiple don't have this though...
+                    upsertMultipleRequest.RequestUri = new Uri(
+                        upsertMultipleRequest.RequestUri.ToString() + "?tag=UpsertMultiple",
+                        uriKind: UriKind.Relative);
+
+                    if (Settings.BypassCustomPluginExecution)
+                    {
+#pragma warning disable CS0162 // Unreachable code detected: Configurable by setting
+                        upsertMultipleRequest.Headers.Add("MSCRM.BypassCustomPluginExecution", "true");
+#pragma warning restore CS0162 // Unreachable code detected: Configurable by setting
+                    }
+
+
+                    Console.WriteLine($"Sending POST request to /{tableSetName}/Microsoft.Dynamics.CRM.UpsertMultiple");
+
+                    await service.SendAsync(upsertMultipleRequest);
+
+                    // Prepare above created entity records to be updated by UpsertMultiple
+
+                    Console.WriteLine($"\nPreparing {numRecCreate} records to Upsert(update)..");
+
+                    var retrieveMultipleQuery = $"{tableLogicalName}s?$select={tablePrimaryKeyName}";
+
+                    if(Settings.CreateAlternateKey)
+                    {
+                        if (Settings.UseElastic)
+                        {
+                            retrieveMultipleQuery += $",{elasticTablePartitionId}";
+                        }
+                        else
+                        {
+                            retrieveMultipleQuery += $",{tableAlternateKey}";
+                        }
+                    }
+
+                    // Retrieve the record ids for records to be updated using PK
+                    var retrieveMultipleRequest = new RetrieveMultipleRequest(
+                        queryUri: retrieveMultipleQuery);
+
+                    
+                    var retrieveMultipleResponse = await service.SendAsync<RetrieveMultipleResponse>(retrieveMultipleRequest);
+
+                    // TODO: RetrieveMultipleResponse.Records contains a JArray of JObjects
```

Comment:
I'm not clear why you are processing the Content when the RetrieveMultipleResponse class does this for you. A Records property should contain the records.

Let me know if this isn't working for you.